Source/Core/VideoBackends/D3D12/D3DCommandListManager.cpp

```diff
@@ -0,0 +1,352 @@
+	}
+
+	m_current_command_allocator_list = (m_current_command_allocator_list + 1) % ARRAYSIZE(m_command_allocator_lists);
+	// End Command Allocator Resets
+
+	m_queue_rollover_fence_value++;
+#ifdef USE_D3D12_QUEUED_COMMAND_LISTS
+	m_queued_command_list->QueueFenceGpuSignal(m_queue_rollover_fence, m_queue_rollover_fence_value);
+#else
+	CheckHR(m_command_queue->Signal(m_queue_rollover_fence, m_queue_rollover_fence_value));
+#endif
+
+	if (m_current_command_allocator_list == 0)
+	{
+		D3D::MoveToNextD3DTextureUploadHeap();
+	}
+
+	// Shader constant buffers are rolled-over independently of command allocator usage.
+	ShaderConstantsManager::CheckToResetIndexPositionInUploadHeaps();
+}
+
+void D3DCommandListManager::ResetCommandListWithIdleCommandAllocator()
+{
+#ifdef USE_D3D12_QUEUED_COMMAND_LISTS
+	ID3D12QueuedCommandList* command_list = m_queued_command_list;
+#else
+	ID3D12GraphicsCommandList* command_list = m_backing_command_list;
+#endif
+
+	CheckHR(command_list->Reset(m_command_allocator_lists[m_current_command_allocator_list][m_current_command_allocator], nullptr));
+}
+
+void D3DCommandListManager::DestroyResourceAfterCurrentCommandListExecuted(ID3D12Resource* resource)
+{
+	CHECK(resource, "Null resource being inserted!");
+
+	m_deferred_destruction_lists[m_current_deferred_destruction_list].push_back(resource);
+}
+
+void D3DCommandListManager::ImmediatelyDestroyAllResourcesScheduledForDestruction()
+{
+	for (UINT i = 0; i < ARRAYSIZE(m_deferred_destruction_lists); i++)
+	{
+		for (UINT j = 0; j < m_deferred_destruction_lists[i].size(); j++)
+		{
+			m_deferred_destruction_lists[i][j]->Release();
+		}
+
+		m_deferred_destruction_lists[i].clear();
+	}
+}
+
+void D3DCommandListManager::ClearQueueAndWaitForCompletionOfInflightWork()
+{
+	// Wait for GPU to finish all outstanding work.
+	m_queue_fence_value++;
+#ifdef USE_D3D12_QUEUED_COMMAND_LISTS
+	m_queued_command_list->ClearQueue(); // Waits for currently-processing work to finish, then clears queue.
+	m_queued_command_list->QueueFenceGpuSignal(m_queue_fence, m_queue_fence_value);
+	m_queued_command_list->ProcessQueuedItems();
+#else
+	CheckHR(m_command_queue->Signal(m_queue_fence, m_queue_fence_value));
+#endif
+	WaitOnCPUForFence(m_queue_fence, m_queue_fence_value);
+}
+
+D3DCommandListManager::~D3DCommandListManager()
+{
+	ImmediatelyDestroyAllResourcesScheduledForDestruction();
+
+#ifdef USE_D3D12_QUEUED_COMMAND_LISTS
+	m_queued_command_list->Release();
+#endif
+	m_backing_command_list->Release();
+
+	for (unsigned int i = 0; i < ARRAYSIZE(m_command_allocator_lists); i++)
+	{
+		for (unsigned int j = 0; j < m_command_allocator_lists[i].size(); j++)
+		{
+			m_command_allocator_lists[i][j]->Release();
+		}
+	}
+
+	m_queue_fence->Release();
+	m_queue_rollover_fence->Release();
+}
+
+void D3DCommandListManager::WaitOnCPUForFence(ID3D12Fence* fence, UINT64 fence_value)
+{
+	HANDLE fence_event = CreateEvent(NULL, FALSE, FALSE, NULL);
```

Comment:
Creating a new event every time we wait for the GPU is unnecessary, re-use the event object (I know the DX samples are written this way, but there's no reason why you can't re-use the handle)